File: cli/src/commands/session/terminate.ts
import { Command } from "commander";
import { daemonDelete } from "../../lib/daemon-client.js";
import { preflight } from "../../lib/preflight.js";
import { getVSTSession } from "../../lib/env.js";
import { die, success } from "../../lib/output.js";

export function registerSessionTerminate(session: Command): void {
  session
    .command("terminate [id]")
    .description("Terminate a session (defaults to your own session, $VST_SESSION, if no id is given)")
    .action(async (id: string | undefined) => {
      const targetId = id ?? getVSTSession();
      if (!targetId) {
        die("No session id given and $VST_SESSION is not set — pass an id explicitly.");
      }

      await preflight();

      const result = await daemonDelete<void>(`/sessions/${targetId}`);

      if (!result.ok) {
        die(result.error, result.status === 404 ? 2 : 1);
      }

      success(`Session terminated: ${targetId}`);
    });
}

File: rust/vst-cli/src/commands/session/terminate.rs
use crate::client::{daemon_delete, DaemonResult};
use crate::env::get_vst_session;
use crate::output::success;
use crate::preflight::preflight;

pub async fn run_session_terminate(id: Option<String>) -> Result<(), (String, i32)> {
    let target_id = id.or_else(get_vst_session);
    let target_id = match target_id {
        Some(t) if !t.is_empty() => t,
        _ => {
            return Err((
                "No session id given and $VST_SESSION is not set — pass an id explicitly."
                    .to_string(),
                1,
            ))
        }
    };

    preflight().await;

    let result = daemon_delete::<serde_json::Value>(&format!("/sessions/{target_id}"))
        .await
        .map_err(|e| (e.to_string(), 1))?;

    match result {
        DaemonResult::Ok { .. } => {
            success(&format!("Session terminated: {target_id}"));
            Ok(())
        }
        DaemonResult::Err { status, error, .. } => {
            let code = if status == 404 { 2 } else { 1 };
            Err((error, code))
        }
    }
}
